kernel/kernel-impl/src/main/java/org/sakaiproject/authz/impl/SakaiSecurity.java

```diff
@@ -705,6 +725,74 @@ protected boolean checkAuthzGroups(String userId, String function, String entity
 			azgs = ref.getAuthzGroups(userId);
 		}
 
+		// need to know whether role swap is in effect, since we can't share the cache entry between sessions
+		// that are swapped and not swapped
+
+		String roleswap = null; // define the roleswap variable
+
+		// this version seems to be enough
+		if (azgs != null && userId != null && userId.equals(sessionManager().getCurrentSessionUserId())) {
```

Comment:
You need to check all refs to see if a user is role swapped i.e. "/assignemnts/". The only way to is to use the entitymanager to get all authzgroups and find the one that starts with site.